Add account_mptokens and account_mptoken_issuances Clio methods (#3754)
XRPLF/xrpl-dev-portal: docs/references/http-websocket-apis/public-api-methods/clio-methods/account_mptoken_issuances.md

---
seo:
    description: Get the MPTokenIssuances created by a given account as of a given ledger.
labels:
  - Accounts
  - Multi Purpose Tokens, MPTs
---

# account_mptoken_issuances

[[Source]](https://github.com/XRPLF/clio/blob/develop/src/rpc/handlers/AccountMPTokenIssuances.cpp "Source")

> The `account_mptoken_issuances` method returns information about the [MPTokenIssuance entries][MPTokenIssuance entry] created by a given account. This method may return large data sets, so you should expect to implement paging via the `marker` field. This API is only available using Clio, not `xrpld`. {% badge href="https://github.com/XRPLF/clio/releases/tag/2.7.0" %}New in: Clio v2.7.0{% /badge %}

Reviewer: Similar feedback as on the other method:

- Backlink, not deep link, in the intro paragraph
- Link explanation for pagination
- Move Clio-only sentence to separate paragraph

```suggestion
The `account_mptoken_issuances` method returns information about [MPT](/docs/concepts/tokens/fungible-tokens/multi-purpose-tokens) issuances created by an account. This method may return large data sets, so you should expect to implement [pagination][Marker]. {% badge href="https://github.com/XRPLF/clio/releases/tag/2.7.0" %}New in: Clio v2.7.0{% /badge %}

This API is only available using Clio, not `xrpld`.
```

## Request Format

{% tabs %}

{% tab label="WebSocket" %}
```json
{
  "command": "account_mptoken_issuances",
  "account": "rExB3uFDE92wKYP6vfjt31Lbi6tXpFobUT",
  "ledger_index": "validated"
}
```
{% /tab %}

{% tab label="JSON-RPC" %}
```json
{
  "method": "account_mptoken_issuances",
  "params": [
    {
      "account": "rExB3uFDE92wKYP6vfjt31Lbi6tXpFobUT",
      "ledger_index": "validated"
    }
  ]
}
```
{% /tab %}

{% /tabs %}

{% try-it method="account_mptoken_issuances" /%}

The request contains the following parameters:

| Field          | Type                      | Required? | Description |
|:---------------|:--------------------------|:----------|-------------|

> | `account`      | String                    | Yes       | The unique identifier of an account, typically the account's address. |

Reviewer: See my comment on `account_mptokens.md`:

```suggestion
| `account`      | String - [Address][]      | Yes       | Look up MPT issuances created by this account. |
```

> | `ledger_index` | [Ledger Index][]          | No        | The [Ledger Index][] of the max ledger to use, or a shortcut string to choose a ledger automatically. See [Specifying Ledgers][]. |

Reviewer: Is "max" a typo here too?

```suggestion
| `ledger_index` | [Ledger Index][]          | No        | The [Ledger Index][] of the ledger version to use, or a shortcut string to choose a ledger automatically. See [Specifying Ledgers][]. |
```

| `ledger_hash`  | String                    | No        | A 32-byte hex string for the ledger version to use. See [Specifying Ledgers][]. |
| `marker`       | [Marker][]                | No        | Used to continue your query where it left off in paginating. |
| `limit`        | Number (positive integer) | No        | Specify a limit to the number of MPTokenIssuances returned. |

## Response Format

{% tabs %}
{% tab label="WebSocket" %}
{% code-snippet file="/_api-examples/account_mptoken_issuances/ws-response.json" language="json" /%}
{% /tab %}

{% tab label="JSON-RPC" %}
{% code-snippet file="/_api-examples/account_mptoken_issuances/jsonrpc-response.json" language="json" prefix="200 OK\n\n" /%}
{% /tab %}

{% /tabs %}

The response follows the [standard format][], with the result containing the following fields:

| Field           | Type             | Description |
|:----------------|:-----------------|:------------|

> | `account`       | String           | The address of the account whose MPTokenIssuances were queried. |

Reviewer: ```suggestion
| `account`       | String           | The account whose MPT issuances were queried. |
```

| `mpt_issuances` | Array            | An array of `MPTokenIssuance` objects created by the account. |
| `marker`        | [Marker][]       | Used to continue querying where we left off when paginating. Omitted if there are no more entries after this result. |
| `limit`         | Number           | The limit, as specified in the request. |
| `ledger_hash`   | String           | The hash of the ledger version used to generate this response. |
| `ledger_index`  | [Ledger Index][] | The index of the ledger version used to generate this response. |
| `validated`     | Boolean          | If `true`, the ledger has been validated by the consensus process and is immutable. Otherwise, the contents of the ledger are not final and may change. In Clio, this is _always_ true as Clio stores and returns validated ledger data. |

> #### MPTokenIssuance

Reviewer: Why h4 and not h3 here? I think the other page has the corresponding data structure as h3.

Similar point to the other, maybe give this synthetic object type a different name to distinguish it from the canonical `MPTokenIssuance` ledger entry type.

Each `MPTokenIssuance` object has the following fields. Fields marked _(May be omitted)_ are only present when they have a meaningful (non-default) value.

| Field                | Type    | Description |
|:---------------------|:--------|:------------|
| `mpt_issuance_id`    | String  | The `MPTokenIssuanceID` of this issuance. |

> | `issuer`             | String  | The account address that created this issuance. |

Reviewer: ```suggestion
| `issuer`             | String  | The account that created this issuance. |
```

| `sequence`           | Number  | The sequence number of the transaction that created this issuance. |

> | `transfer_fee`       | Number  | _(May be omitted)_ The fee charged, in units of 1/100,000, when tokens of this issuance are transferred between non-issuer accounts. |

Reviewer: If omitted, does that mean no transfer fee? If that's right, say so.

| `asset_scale`        | Number  | _(May be omitted)_ The number of decimal places used to represent amounts of this token. |
| `maximum_amount`     | Number  | _(May be omitted)_ The maximum number of tokens that can be issued for this `MPTokenIssuance`. |
| `outstanding_amount` | Number  | _(May be omitted)_ The number of tokens of this issuance currently held by all holders. |
| `locked_amount`      | Number  | _(May be omitted)_ The number of tokens of this issuance currently held in escrow or otherwise locked. |
| `mptoken_metadata`   | String  | _(May be omitted)_ Arbitrary metadata about this issuance, in hexadecimal. |
| `domain_id`          | String  | _(May be omitted)_ The `PermissionedDomainID` associated with this issuance, if any. |

> The following boolean capability flags are included only when they are `true`:

Reviewer: I'm not totally sure it makes sense to separate these out. It might make them easier to skim over by accident.

| Field                 | Type    | Description |
|:----------------------|:--------|:------------|

> | `mpt_locked`          | Boolean | All balances of this issuance are locked. |

Reviewer: It would be good to link the concept article here, but I guess there isn't one for MPT locking. That's unfortunate.

I guess #3337 is the ticket for that. No action needed in this PR, but it would be good to update this line with a link to that concept article when it's ready.

| `mpt_can_lock`        | Boolean | The issuer can lock individual balances or the entire issuance. |

> | `mpt_require_auth`    | Boolean | Individual holders must be authorized by the issuer before they can hold this token. |

Reviewer: It's too bad we don't have a concept article for allow-listing MPTs to link here, although we do at least have issue #3392 on the backlog to add one. This would be a good place to update when that issue is done.

(No action needed here as part of this PR, but flagging it so that issue can have a note about updating this to link to the concept article when it exists.)

> | `mpt_can_escrow`      | Boolean | Holders can place balances of this token into escrow. |

Reviewer: This is another place where you can usefully backlink the concept article, which actually does already exist in this case.

```suggestion
| `mpt_can_escrow`      | Boolean | Holders can place balances of this token into [escrow](/docs/concepts/payment-types/escrow). |
```

> | `mpt_can_trade`       | Boolean | Holders can trade balances of this token on the decentralized exchange. |

Reviewer: Not sure which article is best to link here. But given that actually trading MPTs on the DEX requires another amendment which isn't (currently) even released, it's probably worth a link and/or an `{% amendment-disclaimer ... /%}`. (Currently I believe the amendment that's expected to add this functionality is `MPTokensV2`.)

Author: Yeah MPT DEX is still in development. I've added a link to the amendment page which I think should help for now.

| `mpt_can_transfer`    | Boolean | Tokens of this issuance can be transferred between non-issuer accounts. |

> | `mpt_can_clawback`    | Boolean | The issuer can claw back tokens of this issuance from holders. |

Reviewer: ```suggestion
| `mpt_can_clawback`    | Boolean | The issuer can [claw back tokens](/docs/concepts/tokens/fungible-tokens/clawing-back-tokens) of this issuance from holders. |
```

  "params": [
    {
      "account": "rExB3uFDE92wKYP6vfjt31Lbi6tXpFobUT",
      "ledger_index": "validated"
    }
  ]
}
```
{% /tab %}

{% /tabs %}

{% try-it method="account_mptoken_issuances" /%}

The request contains the following parameters:

| Field          | Type                      | Required? | Description |
|:---------------|:--------------------------|:----------|-------------|
| `account`      | String                    | Yes       | The unique identifier of an account, typically the account's address. |
| `ledger_index` | [Ledger Index][]          | No        | The [Ledger Index][] of the max ledger to use, or a shortcut string to choose a ledger automatically. See [Specifying Ledgers][]. |
| `ledger_hash`  | String                    | No        | A 32-byte hex string for the ledger version to use. See [Specifying Ledgers][]. |
| `marker`       | [Marker][]                | No        | Used to continue your query where it left off in paginating. |
| `limit`        | Number (positive integer) | No        | Specify a limit to the number of MPTokenIssuances returned. |

## Response Format

{% tabs %}
{% tab label="WebSocket" %}

> ```json

Reviewer: Rather than inlining long examples like these, I'd prefer to put JSON files in the `_api-examples/` and use `{% code-snippet ... /%}` to include them here. It makes the MD files a little easier to read, reduces redundant content in translation files, and makes it easier to verify & update the example responses later. A lot of legacy docs still inline them, but you'll also see a good number of recent examples that use the code-snippet pattern.
